apollo-router/src/services/subgraph_service.rs

```diff
@@ -110,151 +138,258 @@ impl tower::Service<crate::SubgraphRequest> for SubgraphService {
+                        ),
+                    }))
+                } else {
+                    Err(BoxError::from(FetchError::SubrequestHttpError {
+                        service: service_name.clone(),
+                        reason: format!("subgraph didn't return JSON (expected content-type: {} or content-type: {GRAPHQL_JSON_RESPONSE_HEADER_VALUE}; found content-type: {content_type:?})", APPLICATION_JSON.essence_str()),
+                    }))
+                };
+            }
+        }
+    }
+
+    let body = hyper::body::to_bytes(body)
+        .instrument(tracing::debug_span!("aggregate_response_data"))
+        .await
+        .map_err(|err| {
+            tracing::error!(fetch_error = format!("{:?}", err).as_str());
 
-            if display_body {
-                tracing::info!(
-                    http.response.body = %String::from_utf8_lossy(&body), apollo.subgraph.name = %service_name, "Raw response body from subgraph {service_name:?} received"
-                );
+            FetchError::SubrequestHttpError {
+                service: service_name.clone(),
+                reason: err.to_string(),
             }
+        })?;
 
-            let graphql: graphql::Response = tracing::debug_span!("parse_subgraph_response")
-                .in_scope(|| {
-                    graphql::Response::from_bytes(&service_name, body).map_err(|error| {
-                        FetchError::SubrequestMalformedResponse {
-                            service: service_name.clone(),
-                            reason: error.to_string(),
-                        }
-                    })
-                })?;
+    if display_body {
+        tracing::info!(
+            http.response.body = %String::from_utf8_lossy(&body), apollo.subgraph.name = %service_name, "Raw response body from subgraph {service_name:?} received"
+        );
+    }
+
+    let graphql: graphql::Response =
+        tracing::debug_span!("parse_subgraph_response").in_scope(|| {
+            graphql::Response::from_bytes(&service_name, body).map_err(|error| {
+                FetchError::SubrequestMalformedResponse {
+                    service: service_name.clone(),
+                    reason: error.to_string(),
+                }
+            })
+        })?;
 
-            let resp = http::Response::from_parts(parts, graphql);
+    let resp = http::Response::from_parts(parts, graphql);
 
-            Ok(crate::SubgraphResponse::new_from_response(resp, context))
-        })
+    Ok(crate::SubgraphResponse::new_from_response(resp, context))
+}
+
+fn get_apq_error(response: graphql::Response) -> APQError {
+    let not_found_byte_string = &Value::String(ByteString::from(PERSISTED_QUERY_NOT_FOUND_ERR_STRING));
+    let not_supported_byte_string = &Value::String(ByteString::from(PERSISTED_QUERY_NOT_SUPPORTED_ERR_STRING));
+    for error in response.errors {
+        match error.extensions.get(&ByteString::from(CODE_STRING)) {
```

Comment:
also add a check for error.message = "PersistedQueryNotFound" since not every server seems to send the code in extensions like [this one](https://github.com/graphql-java/graphql-java/blob/fec817518772e0bcc1cdce1ee10193fe659973fd/src/main/java/graphql/execution/preparsed/persisted/PersistedQueryNotFound.java#L34)